File: command/command.test.ts
import { AbcBank } from "./receiver/abc_bank";
import { XyzBank } from "./receiver/xyz_bank";
import { Transfer } from "./transfer";

test('Command Run', () => {
    const arrayLikeQueue = new Array<Transfer>();

    const abcBank = new AbcBank();
    const transferToAbcBank = new Transfer(transferParam => abcBank.transfer(transferParam));
    arrayLikeQueue.push(transferToAbcBank);

    const xyzBank = new XyzBank();
    const transferToXyzBank = new Transfer(transferParam => xyzBank.transfer(transferParam));
    arrayLikeQueue.push(transferToXyzBank);

    while (arrayLikeQueue.length > 0) {
        const transfer = arrayLikeQueue.shift();
        transfer.start({
            amount: 123.45,
            account: '012345',
        })
    }
});

File: go/command/command_test.go
package command

import "testing"

func TestCommand(t *testing.T) {
	t.Run("Command Run", func(t *testing.T) {
		ch := make(chan *transfer, 2)

		abcBank := NewAbcBank()
		transferToAbcBank := NewTransfer(func(transferParam TransferParam) {
			err := abcBank.Transfer(transferParam)
			if err != nil {
				t.Error(err)
			}
		})
		ch <- transferToAbcBank

		xyzBank := NewXyzBank()
		transferToXyzBank := NewTransfer(func(transferParam TransferParam) {
			xyzBank.Transfer(transferParam)
		})
		ch <- transferToXyzBank

		close(ch)

		for transfer := range ch {
			transfer.Start(NewTransferParam(123.45, "012345"))
		}
	})
}
